In [13]:
from uploading import UploadData
from data_preparing import DataPreprocessing

# О датасете MovieLens100k

In [4]:
movilens_ratings_100k = UploadData.movielens_100k_df()
movilens_ratings_100k

UserId  MovieId  Rating  Timestamp
0         196      242       3  881250949
1         186      302       3  891717742
2          22      377       1  878887116
3         244       51       2  880606923
4         166      346       1  886397596
...       ...      ...     ...        ...
99995     880      476       3  880175444
99996     716      204       5  879795543
99997     276     1090       1  874795795
99998      13      225       2  882399156
99999      12      203       3  879959583

[100000 rows x 4 columns]

In [5]:
users_count =movilens_ratings_100k['UserId'].nunique()
items_count = movilens_ratings_100k['MovieId'].nunique()
print(f"users = {users_count}; items = {items_count}")

users = 943; items = 1682


Сколько записей есть о каждом пользователе и фильме:

In [9]:
users_gr = movilens_ratings_100k.value_counts("UserId")
item_gr = movilens_ratings_100k.value_counts("MovieId")

In [10]:
users_gr.describe()

count    943.000000
mean     106.044539
std      100.931743
min       20.000000
25%       33.000000
50%       65.000000
75%      148.000000
max      737.000000
dtype: float64

In [11]:
item_gr.describe()

count    1682.000000
mean       59.453032
std        80.383846
min         1.000000
25%         6.000000
50%        27.000000
75%        80.000000
max       583.000000
dtype: float64

Что видим: о каждом пользователе есть 20 записей, среднее количество записей чуть больше 100, квантиль 0.75 равна 148. Посмотрим, сколько для каждого пользователя будет доступно записей для обучения:

In [14]:
movielens_preprocessor = DataPreprocessing("UserId", "MovieId", "Rating", "Timestamp", 1, 5)

In [34]:
train_df, test_df = movielens_preprocessor._train_test_split_per(movilens_ratings_100k, 0.2)

In [28]:
users_cou = train_df.value_counts("UserId")
users_cou.describe()

count    943.000000
mean      85.224814
std       80.757562
min       16.000000
25%       27.000000
50%       52.000000
75%      119.000000
max      590.000000
dtype: float64

Также получаем, что не все элементы(фильмы) попали и в тестовый и в тренировочный датасет (логично, ведь для некоторых из них есть всего 1-2 записи в данных)

In [29]:
train_df.value_counts("MovieId").describe()

count    1615.000000
mean       49.762848
std        71.098106
min         1.000000
25%         5.000000
50%        20.000000
75%        65.000000
max       521.000000
dtype: float64

In [20]:
test_df.value_counts("MovieId").describe()

count    1498.000000
mean       13.106142
std        12.476782
min         1.000000
25%         3.000000
50%         9.000000
75%        19.000000
max        69.000000
dtype: float64

# О датасете MovieLens1m

In [23]:
movilens_ratings_1m = UploadData.movielens_1m_df()
movilens_ratings_1m

/home/aleksandra/.local/lib/python3.8/site-packages/pandas/util/_decorators.py:311: ParserWarning: Falling back to the 'python' engine because the 'c' engine does not support regex separators (separators > 1 char and different from '\s+' are interpreted as regex); you can avoid this warning by specifying engine='python'.
  return func(*args, **kwargs)


UserId  MovieId  Rating  Timestamp
0             1     1193       5  978300760
1             1      661       3  978302109
2             1      914       3  978301968
3             1     3408       4  978300275
4             1     2355       5  978824291
...         ...      ...     ...        ...
1000204    6040     1091       1  956716541
1000205    6040     1094       5  956704887
1000206    6040      562       5  956704746
1000207    6040     1096       4  956715648
1000208    6040     1097       4  956715569

[1000209 rows x 4 columns]

In [24]:
users_count =movilens_ratings_1m['UserId'].nunique()
items_count = movilens_ratings_1m['MovieId'].nunique()
print(f"users = {users_count}; items = {items_count}")

users = 6040; items = 3706


In [30]:
users_gr = movilens_ratings_1m.value_counts("UserId")
item_gr = movilens_ratings_1m.value_counts("MovieId")

In [31]:
users_gr.describe()

count    6040.000000
mean      165.597517
std       192.747029
min        20.000000
25%        44.000000
50%        96.000000
75%       208.000000
max      2314.000000
dtype: float64

In [32]:
item_gr.describe()

count    3706.000000
mean      269.889099
std       384.047838
min         1.000000
25%        33.000000
50%       123.500000
75%       350.000000
max      3428.000000
dtype: float64

Посмотрим, сколько в данном датасете будет доступно записей в среднем для каждого пользователя

In [36]:
train_df, test_df = movielens_preprocessor._train_test_split_per(movilens_ratings_1m, 0.2)

In [37]:
train_df.value_counts("UserId").describe()

count    6040.000000
mean      132.873013
std       154.208278
min        16.000000
25%        36.000000
50%        77.000000
75%       167.000000
max      1852.000000
dtype: float64

Также, как и в датасете 100к, не все фильмы попадут и в тренировочный, и в тестовый датасет

In [38]:
train_df.value_counts("MovieId").describe()

count    3667.000000
mean      218.858195
std       329.917870
min         1.000000
25%        25.000000
50%        92.000000
75%       273.500000
max      3167.000000
dtype: float64

In [40]:
test_df.value_counts("MovieId").describe()

count    3534.000000
mean       55.929825
std        60.676376
min         1.000000
25%        11.000000
50%        33.000000
75%        82.000000
max       461.000000
dtype: float64